Automatically sanitize updated code (style fixer)

dantleech opened the pull request:
Introduces a basic "style" fixer mechanism which can be applied on generated / updated code.

Ultimately the "fixers" should be configurable to any given code-style, although the aim is to provide sensible code, not perfectly code (use actual cs fixers for this).

Initially it should support:

- [x] Indentation.
- [x] Empty lines around class members.
- [x] Docblock fixer

As a follow up we should remove all the horrible formatting concerns from the updating code.

dantleech commented on lib/Adapter/TolerantParser/Fixer/IndentationFixer.php:
```diff
@@ -0,0 +1,116 @@
+// 1 Indent the area between the node start and the first child
+// 2 Indent the area between the last child end and the node end
+// 3 If node is structural, level ++
+// 4 Iterate goto 1
```
remove or correct

elythyr commented on lib/Domain/Fixer/DummyFixer.php:
```diff
@@ -0,0 +1,13 @@
+
+use Phpactor\CodeBuilder\Domain\StyleFixer;
+
+class DummyFixer implements StyleFixer
```
```suggestion
class NullFixer implements StyleFixer
```

Personally `DummyFixer` makes me think of a test double, I think `NullFixer` might be more accurate since it respects the general convention around the [null object pattern](https://en.wikipedia.org/wiki/Null_object_pattern)

elythyr commented:
I started the review but there was a lot of little things, then I noticed some case missing for the `MemberEmptyLineFixer` so I tried to implement them as a practice:
https://github.com/elythyr/code-builder/tree/fixer
I let you take a look and take whatever you want from it.

I was wondering mainly about the carriage returns, should we use `PHP_EOL` instead of "\n" ?
For the indentations, instead of counting the number of chars and rebuild it with spaces I just kept the indentation from the user as it is.
Because if they use tabulations then the count is 1 and we will insert only one spaces.
The handling on phpdoc was missing,so we would had an incorrect result when generating properties with a docblock.

dantleech commented:
> I was wondering mainly about the carriage returns, should we use PHP_EOL instead of "\n" ?

No reason in particular, actually not sure if we should allow `\r` and `\r\n`

> The handling on phpdoc was missing,so we would had an incorrect result when generating properties with a docblock.

Yes -- I couldn't figure out how to reasonably handle phpdoc formatting yesterday. Either by adding special case to the indentation fixer, or by having a separate fixer for phpdocs.

dantleech commented:
> I was wondering mainly about the carriage returns, should we use PHP_EOL instead of "\n" ?

Sorry, misunderstood the question. Unix, Mac and Windows all use different new line chars, Unix is `\n`, Mac `\r` and windows `\r\n` IIRC. So we need to handle all of these. And I wonder if we are safe to replace them with `\n`

elythyr commented:
> So we need to handle all of these. And I wonder if we are safe to replace them with \n

I agree that we should handle all of them, and I think I made a mistake in my changes.
We should not look for `PHP_EOL` because even on a windows platform a user might have configure his editor to use `\n` instead.
And I think we should not replace them by `\n`, it's out of the responsibility scope for the fixer in my opinion.
And we might want to inject the type of carriage return to use (like we do with the indent).

dantleech commented:
OK, the fixers now return a collection TextEdits, we then do an intersection with the text edits from the update to only keep the fixes applying to the updated code.

dantleech commented on lib/Util/TextFormat.php:
```diff
@@ -18,9 +18,25 @@ public function indent(string $string, int $level = 0)
+    public static function indentationRemove(string $text): string
+    {
+        $text = preg_replace("/\R +/m", "\n", $text);
+        $text = preg_replace("/^ +/m", "", $text);
```
Pretty sure one of these can be removed and the tests still pass...

elythyr replied:
The first one doesn't seem useful, also it replace the carriage return which could be a bad thing.

dantleech commented on lib/Adapter/TolerantParser/Fixer/MemberEmptyLineFixer.php:
```diff
@@ -0,0 +1,167 @@
+    private const META_INDENTATION = 'indentation';
+    private const META_IS_METHOD = 'is_method';
+
+
```
dd

dantleech commented on lib/Adapter/TolerantParser/Fixer/MemberEmptyLineFixer.php:
```diff
@@ -0,0 +1,167 @@
+    private function blankLines(string $string): int
+    {
+        // TODO: Extract this to a tested utility (perhaps stick it in
+        //       text-document package)
```
yes please

elythyr replied:
Rename it as `countBlankLines` and make it return the number of blank lines instead of the number of carriage return ?

dantleech replied:
:+1: it already returns the number of blank lines

dantleech commented on lib/Adapter/TolerantParser/Fixer/MemberEmptyLineFixer.php:
```diff
@@ -0,0 +1,167 @@
+    private function indentation(Node $node)
+    {
+        $whitespace = $node->getLeadingCommentAndWhitespaceText();
+        $newLinePos = strrpos($whitespace, "\n");
```
needs to account for all new line sequences

dantleech commented on lib/Adapter/TolerantParser/Fixer/MemberEmptyLineFixer.php:
```diff
@@ -0,0 +1,167 @@
+        return $edits;
+    }
+
+    private function addBlankLine($edits, $meta)
```
type hints

dantleech commented on lib/Domain/Fixer/ChainFixer.php:
```diff
@@ -0,0 +1,29 @@
+class ChainFixer implements StyleProposer
+{
+    /**
+     * @var StyleFixer[]
```
Invalid type hint

dantleech commented on lib/Adapter/TolerantParser/Fixer/IndentationFixer.php:
```diff
@@ -0,0 +1,108 @@
+            return $edits;
+        }
+
+        $text = substr($text, $start, $length);
```
assuming byte offset?

dantleech commented on lib/Adapter/TolerantParser/Fixer/IndentationFixer.php:
```diff
@@ -0,0 +1,108 @@
+        return $edits;
+    }
+
+    private function getOffsetUntilFirstChild(Node $node)
```
type hint

dantleech commented on lib/Adapter/TolerantParser/Fixer/IndentationFixer.php:
```diff
@@ -0,0 +1,108 @@
+        $text = substr($text, $start, $length);
+
+        // if there are no new lines in the selection, return
+        if (!preg_match('{\R}m', $text)) {
```
Extract to method -- somehow? `if (0 === count(TextUtil::splitLines($text)` ?

dantleech commented:
Updated to progressively fix the text document for each set of changes instead of trying to merge all the proposed text edits from all fixers into one. We filer each set of changes to only apply if they overlap with the text edits made to the document.

Happy enough with this for now. Next step will see how much of the formatting code can be removed from the updater class.

elythyr commented on lib/Domain/TextEdit.php:
```diff
@@ -36,9 +36,9 @@ public static function applyEdits(array $edits, string $text) : string
             $edit = $edits[$i];
 
-            if ($prevEditStart < $edit->start) {
+            if ($prevEditStart < $edit->start || $prevEditStart < $edit->start + $edit->length) {
```
This mean that either `$edit->start` or `$edit->length` can be null ?
Kind of strange, should it even be possible ?

elythyr commented on lib/Domain/TextEdit.php:
```diff
@@ -36,9 +36,9 @@ public static function applyEdits(array $edits, string $text) : string
             $edit = $edits[$i];
 
-            if ($prevEditStart < $edit->start) {
+            if ($prevEditStart < $edit->start || $prevEditStart < $edit->start + $edit->length) {
```
You can remove the first condition actually.

dantleech replied:
This is mostly copy/pasted from the MS tolerant parser library, don't think this condition is even tested, but would rather leave as is (in this PR I removed the second condition, but re-added it laster).

elythyr commented on lib/Adapter/TolerantParser/Fixer/DocblockIndentationFixer.php:
```diff
@@ -0,0 +1,94 @@
+        foreach ($lines as $line) {
+            if (preg_match('{^\s*\*}', $line)) {
+                $line = TextFormat::indentationRemove($line);
+                $line = $baseIndent . TextFormat::indentApply($line, ' ', $level);
```
Isn't it a bit overkill to use `TextFormat` for that given that `$level` actually has a constant value of 1 which will just append the space ?
```suggestion
                $line = $baseIndent .' '. $line;
```

dantleech replied:
Fixed (and remove the `$level`

elythyr commented on lib/Adapter/TolerantParser/Fixer/IndentationFixer.php:
```diff
@@ -0,0 +1,100 @@
+        }
+
+        $text = TextFormat::indentationRemove($text);
+        $text = TextFormat::indentApply($text, $this->indent, $level);
```
I see that for the second time, it might be interesting to add something like `TextFormat::indentReplace`

dantleech replied:
Done

elythyr commented on lib/Adapter/TolerantParser/Fixer/IndentationFixer.php:
```diff
@@ -0,0 +1,100 @@
+            $edits,
+            $node,
+            $level
+        );
```
Seems like you gave the responsibility the the `indent()` method to add the edit to the list when it should only be to create it.

dantleech replied:
It can add zero or one -- so this avoids a conditional to check for a null. It's not amazing, but it's private.

elythyr commented on lib/Adapter/TolerantParser/Fixer/MemberEmptyLineFixer.php:
```diff
@@ -0,0 +1,174 @@
+                $meta[self::META_FIRST] = true;
+            }
+
+            if ($previousNode && $previousNode === $meta[self::META_NODE_CLASS]) {
```
```suggestion
            if ($previousNode === $meta[self::META_NODE_CLASS]) {
```

I would also rename it to `$previousNodeClass` to be perfectly clear about what it is.

elythyr commented on lib/Util/TextFormat.php:
```diff
@@ -16,11 +16,27 @@ public function __construct(string $indentation = '    ')
+
+    public static function indentationRemove(string $text): string
+    {
+        $text = preg_replace("/^ +/m", "", $text);
```
That won't remove tabulation, does it ?
We should handle all vertical spaces here don't you think ?

dantleech replied:
Fixed to understand tabs

dantleech commented:
Updated. I changed al the methods on TextFormat to be instance methods, inject that into the formatters and add the indentation and new lines as dendencies to that.

elythyr commented on lib/Adapter/TolerantParser/Fixer/DocblockIndentationFixer.php:
```diff
@@ -0,0 +1,90 @@
+        return new TextEdit(
+            $node->getFullStart(),
+            $node->getStart() - $node->getFullStart(),
+            $this->textFormat->implodeLines($newLines)
```
Can you also add a test for properties with phpdoc ?
Without phpdoc there must be no empty line between them, with there must be one.
It will be a tricky one to handle, I would say that there must always be a new line before a docblock:
```suggestion
            $this->textFormat->newLineChar() . $this->textFormat->implodeLines($newLines)
```

At the exception of the first node after the class opening brace...

dantleech replied:
We should probably add it in the MemberEmptyLineFixer

dantleech replied:
Updated. It's handled in the MemberEmptyLineFixer

dantleech commented:
Added a new failing test case

dantleech commented on lib/Adapter/TolerantParser/Updater/AbstractMethodUpdater.php:
```diff
@@ -87,19 +87,11 @@ public function updateMethods(Edits $edits, ClassLikePrototype $classPrototype,
-
-            if (false === $classPrototype->methods()->isLast($methodPrototype)) {
-                $edits->after($lastMember, PHP_EOL);
-            }
```
So we could delete 8 lnes of code ... (hopefully more later)

dantleech commented on lib/Domain/StyleFixer.php:
```diff
@@ -0,0 +1,37 @@
+        return $code;
+    }
+
+    public function fixIntersection(string $code, TextEdits $previouslyAppliedChanges): string
```
Added dedicated method for fixing an "intersectin" - we could add a command to run the above `fix` method on a whole file (don't know why, but useful for debugging at least).

dantleech commented on lib/Domain/StyleFixer.php:
```diff
@@ -0,0 +1,37 @@
+        foreach ($this->propsers as $proposer) {
+            $code = $proposer
+                ->propose($code)
+                ->intersection($previouslyAppliedChanges->appliedTextEdits())
```
In order for this to work we need to have the text edit ranges _after_ they have been applied. `appliedTextEdits` modifies the start offset of the text edits as if they had been applied.

dantleech commented:
I updated the existing updater tests to apply the cs fixers, this uncovered a few more issues. Notably the intersection method didn't work as expected (we needed to have an intersection with the "applied" text edits, not the text edits before application).

:crossed_fingers: this is OK now - it's far too big as a PR :grimacing:

elythyr commented on lib/Adapter/TolerantParser/Fixer/MemberEmptyLineFixer.php:
```diff
@@ -0,0 +1,187 @@
+                self::META_PRECEDING_BLANK_LENGTH => $this->blankLength($node),
+                self::META_INDENTATION => $this->indentation($node),
+                self::META_IS_METHOD => $node instanceof MethodDeclaration,
+                self::META_HAS_DOCBLOCK => TextUtil::hasDocblock($node->getLeadingCommentAndWhitespaceText()),
```
You can retrieve the docblock directly from the `$node` and just check if it's null or not.
Might avoid to add another method in the `TextUtil` class

dantleech replied:
Done -- didn't realise that method existed. It's _potentially_ more expensive then the TextUtil method however as it parses the doc comment (but only as far as saying it is a docblock token, it doesn't parse the tags).

dantleech commented on lib/Adapter/TolerantParser/Fixer/MemberEmptyLineFixer.php:
```diff
@@ -0,0 +1,187 @@
+    {
+        $text = $node->getLeadingCommentAndWhitespaceText();
+
+        if (!preg_match('{^(\s*)}m', $text, $matches)) {
```
Yes it does -- I extracted this to `TextUtil` and added a test.

elythyr replied:
Didn't knew `\s` dealt with tabulation as well, good to know

dantleech commented on lib/Adapter/TolerantParser/Fixer/MemberEmptyLineFixer.php:
```diff
@@ -0,0 +1,180 @@
+
+            if ($previousNodeClass && $previousNodeClass === $meta[self::META_NODE_CLASS]) {
+                $meta[self::META_SUCCESSOR] = true;
+            }
```
could inline these

dantleech commented on lib/Domain/TextEdits.php:
```diff
@@ -0,0 +1,85 @@
+        }
+
+        return self::fromTextEdits($applied);
+    }
```
prob. worth adding some docs for some of these methods, or find better names

dantleech commented on lib/Util/TextUtil.php:
```diff
@@ -0,0 +1,55 @@
+
+namespace Phpactor\CodeBuilder\Util;
+
+class TextUtil
```
final

elythyr commented on lib/Util/TextUtil.php:
```diff
@@ -22,6 +22,15 @@ public static function lineIndentation(string $line): string
+    public static function leadingSpace(string $text): string
+    {
+        if (!preg_match("{^(\s*)}m", $text, $matches)) {
+            return 0;
```
Should return an empty string

dantleech replied:
fixed

dantleech commented:
:santa: :boom:

dantleech commented:
Issues:

dantleech commented:
After spending most of today on trying and failing to fix a simple issue (incorrect indentation of parenthesis) I've decided to revert this.
